### PR TITLE
Resurrect indexHint function.

### LINKED ISSUE #9540: indexHint is broken
CH 20.3.1.2531
```
create table XXXX ( t Int64, f Float64) Engine=MergeTree 
order by t settings index_granularity=128;

insert into XXXX select  number*60, 0 from numbers(100000);

SELECT count()
FROM XXXX
WHERE indexHint(t = 42);

┌─count()─┐
│  100000 │
└─────────┘

drop table XXXX;


create table XXXX ( t Int64, f Float64) Engine=MergeTree 
order by t settings index_granularity=8192;

insert into XXXX select  number*60, 0 from numbers(100000);

SELECT count()
FROM XXXX
WHERE indexHint(t = toDateTime(0));

┌─count()─┐
│  100000 │
└─────────┘

```

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add back `indexHint` function. This is for https://github.com/ClickHouse/ClickHouse/issues/21238 . This reverts https://github.com/ClickHouse/ClickHouse/pull/9542 . This fixes https://github.com/ClickHouse/ClickHouse/issues/9540 .

Detailed description / Documentation draft:

.

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Maybe also add some test case from linked issue?

File: tests/queries/0_stateless/01739_index_hint.reference [hunk 1]
```diff
@@ -0,0 +1,14 @@
+0	0	0
+0	1	0
```
Comment (azat) [inline tests/queries/0_stateless/01739_index_hint.reference:2]: @amosbird but why it returns the first row too for `select * from tbl WHERE indexHint(t = 1) order by t` with `index_granularity=1`?

P.S. it worth converting this test to use `-- { echo }` test hint, see https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01564_test_hint_woes.sql for example

Comment (amosbird) [reply tests/queries/0_stateless/01739_index_hint.reference:2]: > but why it returns the first row too for select * from tbl WHERE indexHint(t = 1) order by t with index_granularity=1

It's how primary index work in clickhouse. The current granule starts with value 1 doesn't mean all values in previous granule < 1

**Comment by amosbird** [comment]
> Maybe also add some test case from linked issue?

Sure. Tests are added.

**Review by nikitamikhaylov — APPROVED** [review]

**Pull request merged**